### PR TITLE
Add HstsMiddleware unit tests

### PR DESCRIPTION
**Pull request by oboukli**
The current tests available for `Microsoft.AspNetCore.HttpsPolicy.HstsMiddleware` are integration test. This PR adds unit tests.

- [X] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [X] You've included unit or integration tests for your change, where applicable.
- [ ] You've included inline docs for your change, where applicable.
- [ ] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=34391) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=34391) before we can accept your contribution.<br/><br/>:x: oboukli [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=34391)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=34391) it.</sub>

File: src/Middleware/HttpsPolicy/test/HstsMiddlewareTests.cs [hunk 1]
```diff
@@ -15,13 +15,152 @@
...
 using Microsoft.Extensions.Logging.Testing;
+using Microsoft.Extensions.Options;
 using Microsoft.Net.Http.Headers;
 using Xunit;
 
 namespace Microsoft.AspNetCore.HttpsPolicy.Tests
 {
     public class HstsMiddlewareTests
     {
+#region Unit tests
```
Comment (Tratcher) [inline src/Middleware/HttpsPolicy/test/HstsMiddlewareTests.cs:26]: No regions please. If the test are meaningfully different then they should go in different files.

File: src/Middleware/HttpsPolicy/test/HstsMiddlewareTests.cs [hunk 2]
```diff
@@ -15,13 +15,152 @@
...
+        [Fact]
+        public async Task Invoke_SetsHstsHeader()
+        {
+            var middleware = new HstsMiddleware(innerHttpContext => Task.CompletedTask, new OptionsWrapper<HstsOptions>(new HstsOptions()));
+
+            var context = new DefaultHttpContext();
+            context.Request.Host = new HostString("example.com");
+            context.Request.Scheme = "https";
+
+            await using (context.Response.Body = new MemoryStream())
```
Comment (Tratcher) [inline src/Middleware/HttpsPolicy/test/HstsMiddlewareTests.cs:54]: Why are you dealing with the response body in these tests? The middleware is unrelated to the response body.

**Review by Tratcher — COMMENTED** [review]
You've duplicated a lot of the tests and it's not clear what value that adds. Are there specific places you'd added new coverage?

The TestServer pattern used in the existing tests is an in-memory system for testing middleware.  It lets us test it in an environment similar to a real app without the network flakiness. Low level unit tests like you've added are not usually required unless there's a specific condition that's hard to test otherwise.

**Comment by Tratcher** [comment]
> [ ] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Please pay attention to this criteria. We use issues to agree on what work is valuable before creating a PR.

File: src/Middleware/HttpsPolicy/test/HstsMiddlewareTests.cs (hunk 1 above)
Comment (oboukli) [reply src/Middleware/HttpsPolicy/test/HstsMiddlewareTests.cs]: Removed

File: src/Middleware/HttpsPolicy/test/HstsMiddlewareTests.cs (hunk 2 above)
Comment (oboukli) [reply src/Middleware/HttpsPolicy/test/HstsMiddlewareTests.cs]: Fixed

**Comment by oboukli** [comment]
Thank you, @Tratcher, for your time and feedback. This PR adds the following three new facts:

```
Ctor_ArgumentNextIsNull_ThrowsArgumentNullException
Ctor_ArgumentOptionsIsNull_ThrowsArgumentNullException
Invoke_SetsHstsHeaderWithDefaultThirtyDaysMaxAge
```

The extra three theories are added for completeness.

Background: I was experimenting with a custom version of the HSTS middleware when I noticed that the integration tests were not being considered by my code coverage report, probably due to tool misconfiguration. However, it was straightforward to add the unit tests, and I'd like to contribute them with this PR.

About opening an issue, I'll be opening one requesting a feature soon, but it's not related to tests.

By the way, while writing the tests, I found that the null check on [line 91](https://github.com/dotnet/aspnetcore/blob/1bc50d2d8623ca93c8e45dfa5a686acb73889103/src/Middleware/HttpsPolicy/src/HstsMiddleware.cs#L91) of `HstsMiddleware` is probably redundant, since [ExcludedHosts](https://github.com/dotnet/aspnetcore/blob/1bc50d2d8623ca93c8e45dfa5a686acb73889103/src/Middleware/HttpsPolicy/src/HstsOptions.cs#L43) is never null, hence it's almost impossible to cover [line 93](https://github.com/dotnet/aspnetcore/blob/1bc50d2d8623ca93c8e45dfa5a686acb73889103/src/Middleware/HttpsPolicy/src/HstsMiddleware.cs#L93) with a test.

**Comment by Tratcher** [comment]
> By the way, while writing the tests, I found that the null check on [line 91](https://github.com/dotnet/aspnetcore/blob/1bc50d2d8623ca93c8e45dfa5a686acb73889103/src/Middleware/HttpsPolicy/src/HstsMiddleware.cs#L91) of `HstsMiddleware` is probably redundant, since [ExcludedHosts](https://github.com/dotnet/aspnetcore/blob/1bc50d2d8623ca93c8e45dfa5a686acb73889103/src/Middleware/HttpsPolicy/src/HstsOptions.cs#L43) is never null, hence it's almost impossible to cover [line 93](https://github.com/dotnet/aspnetcore/blob/1bc50d2d8623ca93c8e45dfa5a686acb73889103/src/Middleware/HttpsPolicy/src/HstsMiddleware.cs#L93) with a test.

Agreed, feel free to remove the check.

**Comment by Tratcher** [comment]
Please remove the new duplicate tests and only include tests that add new coverage. The distinction between the test patterns here is not significant enough to warrant the duplication.

**Comment by oboukli** [comment]
Removed the redundant null check, and removed the new duplicate tests.

File: src/Middleware/HttpsPolicy/test/HstsMiddlewareTests.cs [hunk 3]
```diff
@@ -15,13 +15,46 @@
...
+        public void Ctor_ArgumentOptionsIsNull_ThrowsArgumentNullException()
+        {
+            Assert.Throws<ArgumentNullException>(() =>
+            {
+                _ = new HstsMiddleware(innerHttpContext => Task.CompletedTask, options: null);
+            });
+        }
+
+        [Fact]
+        public async Task Invoke_SetsHstsHeaderWithDefaultThirtyDaysMaxAge()
```
Comment (Tratcher) [inline src/Middleware/HttpsPolicy/test/HstsMiddlewareTests.cs:45]: How is this test different from the following SetOptionsWithDefault_SetsMaxAgeToCorrectValue? The only difference I can see is the port.

Comment (oboukli) [reply src/Middleware/HttpsPolicy/test/HstsMiddlewareTests.cs:45]: Shall I remove it?

Comment (Tratcher) [reply src/Middleware/HttpsPolicy/test/HstsMiddlewareTests.cs:45]: Yes if it's not providing unique coverage

Comment (oboukli) [reply src/Middleware/HttpsPolicy/test/HstsMiddlewareTests.cs]: Removed

**Review by Tratcher — APPROVED** [review]

File: src/Middleware/HttpsPolicy/test/HstsMiddlewareTests.cs (hunk 3 above)
Comment (Tratcher) [reply src/Middleware/HttpsPolicy/test/HstsMiddlewareTests.cs]: Thanks

**Pull request merged**